File: ios/JustSpent/JustSpent/Models/Currency.swift
//
//  Currency.swift
//  JustSpent
//
//  Created by Claude Code on 2025-10-19.
//  Comprehensive currency model with metadata and utilities
//  Now loads from shared currencies.json for easy updates
//

import Foundation

/// Represents a currency in the Just Spent application
/// Loaded dynamically from currencies.json for flexibility and easy updates
struct Currency: Codable, Identifiable, Hashable {
    let code: String
    let symbol: String
    let displayName: String
    let shortName: String
    let localeIdentifier: String
    let isRTL: Bool
    let voiceKeywords: [String]

    // MARK: - Identifiable
    var id: String { code }

    // MARK: - Computed Properties

    /// Locale object for formatting operations
    var locale: Locale {
        return Locale(identifier: localeIdentifier)
    }

    /// Number of decimal places for this currency
    var decimalPlaces: Int {
        return 2
    }

    /// Decimal separator character
    var decimalSeparator: String {
        return "."
    }

    /// Thousands grouping separator
    var groupingSeparator: String {
        return ","
    }

    // MARK: - Currency Loading

    /// Cached currencies loaded from JSON
    private static var _allCurrencies: [Currency]?

    /// Initialize the currency system
    /// Should be called once during app initialization
    static func initialize() {
        if _allCurrencies == nil {
            NSLog("🔄 Currency: Starting initialization...")
            _allCurrencies = CurrencyLoader.loadCurrencies()
            NSLog("🔄 Currency: Initialization complete. Loaded %d currencies", _allCurrencies?.count ?? 0)

            if (_allCurrencies?.isEmpty ?? true) {
                NSLog("❌ CRITICAL: Currency.all is EMPTY after initialization!")
                NSLog("❌ This will cause Currency.from() to return nil for all codes")
                NSLog("❌ Multi-currency views will not appear!")
            } else {
                NSLog("✅ Currency system healthy with %d currencies", _allCurrencies!.count)
                NSLog("✅ Sample currencies: %@", _allCurrencies!.prefix(3).map { $0.code }.joined(separator: ", "))
            }
        } else {
            NSLog("⚠️ Currency: Already initialized with %d currencies", _allCurrencies?.count ?? 0)
        }
    }

    /// All supported currencies
    static var all: [Currency] {
        if _allCurrencies == nil {
            initialize()
        }
        return _allCurrencies ?? []
    }

    /// Commonly used currencies (for onboarding UI)
    static var common: [Currency] {
        let commonCodes = ["AED", "USD", "EUR", "GBP", "INR", "SAR", "JPY"]
        return all.filter { commonCodes.contains($0.code) }
    }

    /// Default currency for new users (based on device locale)
    static var `default`: Currency {
        let deviceLocale = Locale.current
        let currencyCode = deviceLocale.currencyCode?.uppercased() ?? "USD"

        // Try to find currency by locale, then USD, then first available
        if let currency = from(isoCode: currencyCode) {
            return currency
        } else if let usd = from(isoCode: "USD") {
            return usd
        } else if let firstCurrency = all.first {
            print("⚠️ Currency: USD not found, using first available currency: \(firstCurrency.code)")
            return firstCurrency
        } else {
            // Fallback for test environment where currencies.json might not load
            print("❌ Currency: No currencies loaded! Using hardcoded AED fallback")
            return Currency(
                code: "AED",
                symbol: "د.إ",
                displayName: "UAE Dirham",
                shortName: "Dirham",
                localeIdentifier: "ar_AE",
                isRTL: true,
                voiceKeywords: ["dirham", "dirhams", "aed"]
            )
        }
    }

    /// Create currency from ISO code string
    /// - Parameter code: ISO 4217 currency code
    /// - Returns: Currency struct or nil if invalid
    static func from(isoCode code: String) -> Currency? {
        return all.first { $0.code.uppercased() == code.uppercased() }
    }

    /// Detect currency from text input (voice commands, manual entry)
    /// - Parameter text: Input text to analyze
    /// - Returns: Detected currency or nil if no match found
    static func detectFromText(_ text: String) -> Currency? {
        let lowercasedText = text.lowercased()

        for currency in all {
            for keyword in currency.voiceKeywords {
                if lowercasedText.contains(keyword.lowercased()) {
                    return currency
                }
            }
        }

        return nil
    }

    // MARK: - Legacy Accessors for Backward Compatibility

    static var aed: Currency { from(isoCode: "AED")! }
    static var usd: Currency { from(isoCode: "USD")! }
    static var eur: Currency { from(isoCode: "EUR")! }
    static var gbp: Currency { from(isoCode: "GBP")! }
    static var inr: Currency { from(isoCode: "INR")! }
    static var sar: Currency { from(isoCode: "SAR")! }
}

// MARK: - CustomStringConvertible

extension Currency: CustomStringConvertible {
    var description: String {
        return "\(symbol) \(code)"
    }
}

// MARK: - Comparable

extension Currency: Comparable {
    static func < (lhs: Currency, rhs: Currency) -> Bool {
        return lhs.displayName < rhs.displayName
    }
}

// MARK: - Internal Loading Mechanism

/// Internal data structure for JSON parsing
private struct CurrenciesData: Codable {
    let version: String
    let lastUpdated: String
    let currencies: [Currency]
}

/// Utility object to load currencies from JSON resource file
private struct CurrencyLoader {
    static func loadCurrencies() -> [Currency] {
        NSLog("📂 Currency: Searching for currencies.json...")

        let fileManager = FileManager.default
        var jsonURL: URL?

        // 1. Try from current working directory and relative paths (works in tests and CI)
        let currentDirPath = fileManager.currentDirectoryPath
        let possiblePaths = [
            "\(currentDirPath)/shared/currencies.json",        // From project root
            "\(currentDirPath)/../shared/currencies.json",     // From ios/ directory
            "\(currentDirPath)/../../shared/currencies.json"   // From ios/JustSpent/ directory
        ]

        for path in possiblePaths {
            if fileManager.fileExists(atPath: path) {
                jsonURL = URL(fileURLWithPath: path)
                NSLog("✅ Currency: Found currencies.json in shared folder (from current dir: %@)", path)
                break
            }
        }

        // 2. Try searching up from bundle path to find project root (works in simulator)
        if jsonURL == nil {
            jsonURL = findCurrencyFileFromBundle(fileManager: fileManager)
        }

        // 3. Fallback: load from bundle (production build with file added to Xcode)
        if jsonURL == nil {
            jsonURL = Bundle.main.url(forResource: "currencies", withExtension: "json")
            if jsonURL != nil {
                NSLog("✅ Currency: Found currencies.json in app bundle")
            }
        }

        guard let url = jsonURL else {
            NSLog("❌ Currency: currencies.json NOT FOUND in any location")
            NSLog("   Searched: shared folder, bundle path, and app bundle")
            NSLog("📂 Currency: Bundle path: %@", Bundle.main.bundlePath)
            NSLog("📂 Currency: Current dir: %@", currentDirPath)
            return []
        }

        NSLog("✅ Currency: Loading from: %@", url.path)

        do {
            let data = try Data(contentsOf: url)
            NSLog("📊 Currency: JSON file size: %d bytes", data.count)

            let decoder = JSONDecoder()
            let currenciesData = try decoder.decode(CurrenciesData.self, from: data)
            NSLog("✅ Currency: Successfully decoded JSON")
            NSLog("✅ Currency: Loaded %d currencies from JSON (version %@)", currenciesData.currencies.count, currenciesData.version)
            NSLog("✅ Currency: Sample currencies: %@", currenciesData.currencies.prefix(3).map { $0.code }.joined(separator: ", "))
            return currenciesData.currencies
        } catch let decodingError as DecodingError {
            NSLog("❌ Currency: JSON DECODING ERROR: %@", String(describing: decodingError))
            switch decodingError {
            case .keyNotFound(let key, let context):
                NSLog("   Missing key: %@ at %@", key.stringValue, String(describing: context.codingPath))
            case .typeMismatch(let type, let context):
                NSLog("   Type mismatch for type: %@ at %@", String(describing: type), String(describing: context.codingPath))
            case .valueNotFound(let type, let context):
                NSLog("   Value not found for type: %@ at %@", String(describing: type), String(describing: context.codingPath))
            case .dataCorrupted(let context):
                NSLog("   Data corrupted at %@: %@", String(describing: context.codingPath), context.debugDescription)
            @unknown default:
                NSLog("   Unknown decoding error")
            }
            return []
        } catch {
            NSLog("❌ Currency: Failed to load currencies from JSON: %@", error.localizedDescription)
            return []
        }
    }

    /// Search for project root by looking for shared/currencies.json
    private static func findCurrencyFileFromBundle(fileManager: FileManager) -> URL? {
        let bundlePath = Bundle.main.bundlePath

        var currentPath = (bundlePath as NSString).deletingLastPathComponent

        // Search up to 10 levels to find project root
        for _ in 0..<10 {
            let sharedPath = (currentPath as NSString).appendingPathComponent("shared/currencies.json")
            if fileManager.fileExists(atPath: sharedPath) {
                print("✅ Currency: Found currencies.json in shared folder (from bundle path)")
                return URL(fileURLWithPath: sharedPath)
            }

            // Go up one level
            let parentPath = (currentPath as NSString).deletingLastPathComponent
            if parentPath == currentPath {
                break // Reached root
            }
            currentPath = parentPath
        }

        return nil
    }
}

File: android/app/src/main/java/com/justspent/expense/data/model/Currency.kt
@file:OptIn(kotlinx.serialization.ExperimentalSerializationApi::class)

package com.justspent.expense.data.model

import android.content.Context
import kotlinx.serialization.ExperimentalSerializationApi
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json
import java.util.Locale

/**
 * Represents a currency in the Just Spent application
 * Loaded dynamically from currencies.json for flexibility and easy updates
 *
 * @property code ISO 4217 currency code
 * @property symbol Currency symbol for display
 * @property displayName Localized currency name in English
 * @property shortName Short display name for compact views
 * @property localeIdentifier Locale identifier for proper number formatting
 * @property isRTL Whether this currency uses right-to-left text direction
 * @property voiceKeywords Keywords for voice command detection
 */
@Serializable
data class Currency(
    val code: String,
    val symbol: String,
    val displayName: String,
    val shortName: String,
    val localeIdentifier: String,
    val isRTL: Boolean,
    val voiceKeywords: List<String>
) {
    /**
     * Locale object for formatting operations
     */
    val locale: Locale by lazy {
        val parts = localeIdentifier.split("_")
        when (parts.size) {
            1 -> Locale(parts[0])
            2 -> Locale(parts[0], parts[1])
            else -> Locale.US
        }
    }

    /**
     * Number of decimal places for this currency
     */
    val decimalPlaces: Int = 2

    /**
     * Decimal separator character
     * Always uses "." (Western/English format) for consistency across all currencies
     */
    val decimalSeparator: String = "."

    /**
     * Thousands grouping separator
     * Always uses "," (Western/English format) for consistency across all currencies
     */
    val groupingSeparator: String = ","

    // MARK: - Overrides

    override fun toString(): String = "$symbol $code"

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is Currency) return false
        return code == other.code
    }

    override fun hashCode(): Int = code.hashCode()

    companion object {
        // Cached currencies loaded from JSON
        private var _allCurrencies: List<Currency>? = null

        /**
         * Initialize the currency system with context
         * Should be called once during app initialization
         */
        fun initialize(context: Context) {
            if (_allCurrencies == null) {
                android.util.Log.d("Currency", "🔄 Starting initialization...")
                _allCurrencies = CurrencyLoader.loadCurrencies(context)
                android.util.Log.d("Currency", "🔄 Initialization complete. Loaded ${_allCurrencies?.size ?: 0} currencies")
            } else {
                android.util.Log.w("Currency", "⚠️ Already initialized with ${_allCurrencies?.size ?: 0} currencies")
            }
        }

        /**
         * All supported currencies
         */
        val all: List<Currency>
            get() = _allCurrencies ?: throw IllegalStateException(
                "Currency system not initialized. Call Currency.initialize(context) first."
            )

        /**
         * Commonly used currencies (for onboarding UI)
         */
        val common: List<Currency>
            get() = all.filter { it.code in commonCodes }

        private val commonCodes = listOf("AED", "USD", "EUR", "GBP", "INR", "SAR", "JPY")

        /**
         * Default currency for new users (based on device locale)
         */
        val default: Currency
            get() {
                val deviceLocale = Locale.getDefault()
                val currencyCode = try {
                    java.util.Currency.getInstance(deviceLocale).currencyCode
                } catch (e: Exception) {
                    "USD"
                }

                return fromCode(currencyCode) ?: fromCode("USD")!!
            }

        /**
         * Create currency from ISO code string
         *
         * @param code ISO 4217 currency code
         * @return Currency object or null if invalid
         */
        fun fromCode(code: String): Currency? {
            return all.find { it.code.equals(code, ignoreCase = true) }
        }

        /**
         * Detect currency from text input (voice commands, manual entry)
         * Prioritizes common currencies (AED, USD, EUR, GBP, INR, SAR) before checking all currencies
         * Uses word boundary matching for better accuracy and prioritizes longer keywords
         *
         * @param text Input text to analyze
         * @return Detected currency or null if no match found
         */
        fun detectFromText(text: String): Currency? {
            val lowercasedText = text.lowercase()

            // Helper function to check if keyword matches (whole word or symbol)
            fun matchesKeyword(text: String, keyword: String): Boolean {
                val lowercaseKeyword = keyword.lowercase()

                // For single character or symbol keywords, use contains
                if (keyword.length <= 2 && keyword.any { !it.isLetterOrDigit() }) {
                    return text.contains(lowercaseKeyword)
                }

                // For word keywords, use word boundary matching
                return text.contains(Regex("\\b${Regex.escape(lowercaseKeyword)}\\b"))
            }

            // Helper function to find best match with tie-breaking logic
            fun findBestMatch(currencies: List<Currency>): Currency? {
                val matches = mutableListOf<Pair<Currency, String>>()

                for (currency in currencies) {
                    // Sort keywords by length (longest first) to prioritize specific phrases
                    // This ensures "australian dollars" is checked before "dollars"
                    val sortedKeywords = currency.voiceKeywords.sortedByDescending { it.length }

                    for (keyword in sortedKeywords) {
                        if (matchesKeyword(lowercasedText, keyword)) {
                            // Only add the first (longest) matching keyword for this currency
                            matches.add(currency to keyword)
                            break  // Stop checking other keywords for this currency
                        }
                    }
                }

                if (matches.isEmpty()) return null

                // Find max keyword length
                val maxLength = matches.maxOf { it.second.length }

                // Get all matches with max length
                val longestMatches = matches.filter { it.second.length == maxLength }

                // If only one longest match, return it
                if (longestMatches.size == 1) {
                    return longestMatches.first().first
                }

                // Tie-breaker: prefer common currencies (USD > AUD for "dollar")
                val commonMatch = longestMatches.firstOrNull { it.first.code in commonCodes }
                if (commonMatch != null) {
                    return commonMatch.first
                }

                // Otherwise return first match
                return longestMatches.first().first
            }

            // Search all currencies with intelligent tie-breaking
            // Prioritizes longest keywords, then common currencies
            return findBestMatch(all)
        }

        // Legacy object references for backward compatibility
        val AED: Currency get() = fromCode("AED")!!
        val USD: Currency get() = fromCode("USD")!!
        val EUR: Currency get() = fromCode("EUR")!!
        val GBP: Currency get() = fromCode("GBP")!!
        val INR: Currency get() = fromCode("INR")!!
        val SAR: Currency get() = fromCode("SAR")!!
    }
}

/**
 * Internal data structure for JSON parsing
 */
@Serializable
private data class CurrenciesData(
    val version: String,
    val lastUpdated: String,
    val currencies: List<Currency>
)

/**
 * Utility object to load currencies from JSON asset file
 */
private object CurrencyLoader {
    private val json = Json { ignoreUnknownKeys = true }

    fun loadCurrencies(context: Context): List<Currency> {
        return try {
            android.util.Log.d("Currency", "📂 Searching for currencies.json in assets...")

            val jsonString = context.assets.open("currencies.json").bufferedReader().use { it.readText() }
            android.util.Log.d("Currency", "✅ Found currencies.json")
            android.util.Log.d("Currency", "📊 JSON file size: ${jsonString.length} characters")

            val data = json.decodeFromString<CurrenciesData>(jsonString)
            android.util.Log.d("Currency", "✅ Successfully decoded JSON")
            android.util.Log.d("Currency", "✅ Loaded ${data.currencies.size} currencies from JSON (version ${data.version})")
            android.util.Log.d("Currency", "✅ Sample currencies: ${data.currencies.take(3).map { it.code }.joinToString(", ")}")

            data.currencies
        } catch (e: kotlinx.serialization.SerializationException) {
            android.util.Log.e("Currency", "❌ JSON SERIALIZATION ERROR", e)
            android.util.Log.e("Currency", "   This usually means the JSON structure doesn't match the Currency data class")
            emptyList()
        } catch (e: java.io.FileNotFoundException) {
            android.util.Log.e("Currency", "❌ currencies.json NOT FOUND in assets folder", e)
            android.util.Log.e("Currency", "   Check if file exists: android/app/src/main/assets/currencies.json")
            emptyList()
        } catch (e: Exception) {
            android.util.Log.e("Currency", "❌ Failed to load currencies from JSON: ${e.javaClass.simpleName}", e)
            android.util.Log.e("Currency", "   Error message: ${e.message}")
            emptyList()
        }
    }
}

/**
 * Extension function to get Currency from String code
 */
fun String.toCurrency(): Currency? = Currency.fromCode(this)
